### PR TITLE
[Web] Beta versioning

### PR DESCRIPTION
**Pull request by jahorton**

### CODE REVIEW

**Comment by keyman-server** [comment]
[OSK renders for Windows](https://build.palaso.org/repository/download/KeymanDesktop_TestPrRenderOnScreenKeyboards/138907:id/kvk/index.html)

File: web/source/kmwkeyboards.ts [hunk 1]
```diff
@@ -227,7 +227,7 @@ namespace com.keyman {
     /**
      *  Create or update a keyboard meta-data 'stub' during keyboard registration
      * 
-     *  Cross-reference with https://help.keyman.com/developer/engine/web/10.0/reference/core/addKeyboards.
+     *  Cross-reference with https://help.keyman.com/developer/engine/web/11.0/reference/core/addKeyboards.
```
Comment (darcywong00) [inline web/source/kmwkeyboards.ts]: ah yes, we'll need to copy developer/engine/web/10.0/ to 11.0

File: web/source/kmwdebug.js [hunk 2]
```diff
@@ -1,6 +1,6 @@
 /***
-   KeymanWeb 10.0
-   Copyright 2017 SIL International
+   KeymanWeb 11.0
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/kmwdebug.js]: I somewhat consciously left it out - it's not included in compilation at the moment, and little's being done to maintain it.  I can update the text if we want that, at least.

**Comment by mcdurdin** [comment]
Is it at all possible to use resources/version.md instead of hardcoding version strings? Or if not, at least add a comment `// see resources/version.md` so we can grep on that for future updates?

**Comment by jahorton** [comment]
@mcdurdin Obviously not for comments, unless we wrote a script to use for major-version transition updates.  This would probably be the 'best' way, considering that NPM also has a slot for our version number and that we're considering publishing KMW to NPM at large.  (A similar script could overwrite the build's package.json with the exact build version string, like we do for the iOS build in its PLists.)

Of course, if we're only worried about the actual version string accessible through KMW and JS, that wouldn't necessarily need a custom script, though there _is_ a simple enough script-based approach that can be used - since TypeScript can `<reference>`-include other source files, we can have a script write an "environment.ts" file with any relevant per-build values for inclusion.  Alternatively, there's the old 'define' through Closure approach, like we currently do for the build number, but TypeScript doesn't seem to have a direct equivalent.

**Comment by darcywong00** [comment]
I like the idea of keeping version numbers in the build scripts and npm package files.

Updating every fileheader for every release seems like more busy work.

**Comment by jahorton** [comment]
By "every release" for fileheaders, do you mean major-version transition, or every build?  The idea's intent is for that script to be used on major-version transitions, not on every build.

**Comment by darcywong00** [comment]
yeah, I meant every major version transition

**Comment by mcdurdin** [comment]
I am not so bothered about comments at this stage. I am looking at resource strings and code references.

> Of course, if we're only worried about the actual version string accessible through KMW and JS, that wouldn't necessarily need a custom script, though there _is_ a simple enough script-based approach that can be used - since TypeScript can `<reference>`-include other source files, we can have a script write an "environment.ts" file with any relevant per-build values for inclusion.

Yes, that sounds like a good way forward. Maybe call it 'environment.inc.ts' or something like that (and add it to .gitignore). It can then be easily built from resources/version.md.

File: web/source/kmwbase.ts [hunk 3]
```diff
@@ -115,7 +115,7 @@ namespace com.keyman {
       this.rootPath = KeymanBase._rootPath;
       this.protocol = KeymanBase._protocol;
 
-      this['version'] = "10.0";
+      this['version'] = "11.0";
```
Comment (mcdurdin) [inline web/source/kmwbase.ts]: Source from `environment.inc.ts`

File: web/source/build.sh [hunk 4]
```diff
@@ -276,6 +278,21 @@ readonly BUILD_DEBUG_EMBED
 readonly BUILD_COREWEB
 readonly DO_MINIFY
 
+generate_environment_ts_file ( ) {
+if [ -f $ENVIRONMENT_FILE ];
```
Comment (mcdurdin) [inline web/source/build.sh]: Can you fix whitespace in this function please (obviously the multiline `echo` contents will still be unindented, but the rest could be better 🙂)

Comment (jahorton) [reply web/source/build.sh]: Wasn't sure which way our conventions would dictate here.  Not a problem.

**Pull request merged**